logging/logging.go

```diff
@@ -1,25 +1,226 @@
 package logging
 
 import (
+	"context"
 	"encoding/json"
 	"io"
+	"sync"
 
+	"github.com/google/uuid"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-func NewLogger(service string) *zap.SugaredLogger {
+type Logger struct {
+	*zap.SugaredLogger
+	id     RequestID
+	Values *sync.Map
+}
+
+type RequestID string
+type contextKey string
+
+const (
+	SkipProviderType string = ""
+	NoVariant        string = ""
+)
+
+const (
+	RequestIDKey = contextKey("request-id")
+	LoggerKey    = contextKey("logger")
+)
+
+func NewRequestID() RequestID {
+	return RequestID(uuid.New().String())
+}
+
+func (r RequestID) String() string {
+	return string(r)
+}
+
+func (logger Logger) WithRequestID(id RequestID) Logger {
+	if id == "" {
+		logger.Warn("Request ID is empty")
+		return logger
+	}
+	if logger.id != "" {
+		logger.Warnw("Request ID already set in logger. Using existing Request ID", "current request-id", logger.id, "new request-id", id)
+		return logger
+	}
+
+	return Logger{SugaredLogger: logger.With("request-id", id),
+		id:     id,
+		Values: logger.Values}
+}
+
+func (logger Logger) WithResource(resourceType, name, variant string) Logger {
```

Comment:
nit: feel like `resourceType` shouldn't be a string but rather `ResourceType`

Reply:
likely to cause a circular dependency